### PR TITLE
DEV 2220 cancel update not pk

### PR DESCRIPTION
**Pull request by x110dc**
Please fill out each section even if it's just with "N/A"

#### What's this PR do?

- moves handling of Subscription cancellation and update out of PaymentManager and into an APIView
- adds a Subscription ID to the contributions serializer 

#### Why are we doing this? How does it help us?

Since we changed to show payments directly from Stripe instead of from Contributions the API needed to be updated to handle updating the payment method and cancelling for Subscriptions that may or may not have an associated Contribution. The Subscription Id was added to the serializer so the API would know the correct object to update/cancel.

#### How should this be manually tested? Please include detailed step-by-step instructions.

Since these changes are only for the API and the frontend will be updated in subsequent ticket this can't be tested with the SPA directly

1. Create a recurring contribution
2. Request a magic link
3. From the results, pull out the JWT, the Subscription ID and the CSRF token and plug them in here:
```
curl -v -F 'revenue_program_slug=billypenn' --header "X-CSRFToken: blah"  -X DELETE --cookie "Authorization=quux" --cookie "csrftoken=blah" http://localhost:8000/api/v1/subscriptions/sub_baz/
```
Then to update the payment method, first create a new one: 
```
stripe payment_methods create --type=card -d "card[number]=4242424242424242" -d "card[exp_month]=8" -d "card[exp_year]=2027" -d "card[cvc]=234" --stripe-account=acct_quux
```

Then update similarly: 
```
 curl -v -F 'revenue_program_slug=billypenn' -F 'payment_method_id=pm_1LavTgPbhdfvD1kVhmGQNdhR' --header "X-CSRFToken: foo" -X PATCH --cookie "Authorization=bar" --cookie "csrftoken=baz" http://localhost:8000/api/v1/subscriptions/sub_swizzle
```

#### Did this PR make changes to the user interface that may require changes to the user-facing docs?

 No.

#### Have automated unit tests been added? If not, why?

Yes.

#### How should this change be communicated to end users?

N/A

#### Are there any smells or added technical debt to note?

- I added more tests in pytest format; so there's a mix of test styles. 
- I wasn't how best to structure the endpoints and verbs; I went with what seemed most RESTful: a DELETE and a PATCH on the bare identifier. 
- Likewise, I wasn't sure what response codes to use. 
- I wasn't sure how to use the DRF permission framework (other than confirming a Contributor) so I just did a bare-simple check of the email against the email for the customer associated with the Subscription. I created a ticket to improve this.
- I wasn't sure about requiring the revenue program as a parameter on the call or inferring it from the subdomain; it looks like both of these methods happen in the code in different places; not sure pros/cons of each but went with the parameter

#### Has this been documented? If so, where?

No.

#### What are the relevant tickets? Add a link to any relevant ones.

https://news-revenue-hub.atlassian.net/browse/DEV-2220

#### Do any changes need to be made before deployment to production (adding environment variables, for example)?

No.

#### Are there next steps? If so, what? Have tickets been opened for them? List next-step tickets here:

- https://news-revenue-hub.atlassian.net/browse/DEV-2287 
- https://news-revenue-hub.atlassian.net/browse/DEV-2286

### CODE REVIEW

File: apps/contributions/serializers.py [hunk 1]
```diff
@@ -474,6 +474,7 @@ class PaymentProviderContributionSerializer(serializers.Serializer):
     # id will be charge object id in our case, which will start with ch_ and doesn't exceed 255 chars
     # https://stripe.com/docs/upgrades#what-changes-does-stripe-consider-to-be-backwards-compatible
     id = serializers.CharField(max_length=255)
+    subscription_id = serializers.CharField(max_length=255, required=False, allow_blank=True)
```
Comment (njharman) [inline apps/contributions/serializers.py:477]: help_text, comment or rename to make it clear this is a Stripe ID and not something NRE related.

File: apps/contributions/stripe_contributions_provider.py [hunk 2]
```diff
@@ -229,6 +235,7 @@ def upsert(self, contributions):
         """Serialized and Upserts contributions data to cache."""
         data = self.serialize(contributions)
         cached_data = json.loads(self.cache.get(self.key) or "{}")
+
```
Comment (njharman) [inline apps/contributions/stripe_contributions_provider.py:238]: boo!

File: apps/contributions/views.py [hunk 3]
```diff
@@ -331,50 +332,71 @@ def process_flagged(self, request, pk=None):
...
+        if request.user.email.lower() != subscription.customer.email.lower():
+            # TODO: [DEV-2287] should we find a way to user DRF's permissioning scheme here instead?
+            return Response({"detail": "Forbidden"}, status=status.HTTP_403_FORBIDDEN)
+        try:
+            stripe.Subscription.delete(
+                subscription_id, stripe_account=revenue_program.payment_provider.stripe_account_id, expand=["customer"]
+            )
+        except stripe.error.StripeError:
+            logger.exception("stripe.Subscription.delete returned a StripeError")
+            return Response({"detail": ""}, status=status.HTTP_500_INTERNAL_SERVER_ERROR)
```
Comment (njharman) [inline apps/contributions/views.py:400]: Should detail be something?

File: apps/contributions/tests/test_views.py [hunk 4]
```diff
@@ -758,75 +760,99 @@ def setUp(self):
...
+            subscription_id=self.subscription_id,
+            data={
+                "payment_method_id": self.payment_method_id,
+                "revenue_program_slug": self.revenue_program.slug,
+            },
+        )
+        assert response.status_code == 403
+        assert response.data["detail"] == "Forbidden"
         mock_modify.assert_not_called()
         mock_attach.assert_not_called()
```
Comment (njharman) [inline apps/contributions/tests/test_views.py:814]: The pytest way would be
> assert not mock_modify.called

File: apps/contributions/tests/test_views.py [hunk 5]
```diff
@@ -758,75 +760,99 @@ def setUp(self):
...
+    @mock.patch("stripe.Subscription.retrieve")
     @mock.patch("stripe.PaymentMethod.attach")
     @mock.patch("stripe.Subscription.modify")
-    def test_failure_when_any_parameter_other_than_pm_id(self, mock_modify, mock_attach):
-        response = self._make_request(self.contribution, {"amount": 10})
+    def test_failure_when_any_parameter_other_than_pm_id(self, mock_modify, mock_attach, mock_retrieve):
+        response = self._make_request(
+            subscription_id=self.subscription_id,
+            data={
+                "fiddle": self.payment_method_id,
```
Comment (njharman) [inline apps/contributions/tests/test_views.py:790]: is "fiddle" well know "bad value"?  I typically use something like "test_unknown_parameter"

**Review by njharman — APPROVED** [review]
lgtm, 100% coverage
Approved with commentary 

> I wasn't sure about requiring the revenue program as a parameter on the call or inferring it from the subdomain; it looks like both of these methods happen in the code in different places; not sure pros/cons of each but went with the parameter

Is the API served from RP subdomains? I thought that was only SPA.  HTTP_REFFERER is not reliable. That is why I required it as param for magic email. 

fwiw 204 is more correct than 200 for DELETE and PATCH (with no content). I believe the Proper PATCH response would be 200 and return the full modified object. But Proper REST is so unpragmatic for APIs (as opppsed to HyperText Documents).

File: apps/contributions/serializers.py (hunk 1 above)
Comment (x110dc) [reply apps/contributions/serializers.py:477]: done

File: apps/contributions/stripe_contributions_provider.py (hunk 2 above)
Comment (x110dc) [reply apps/contributions/stripe_contributions_provider.py]: removed.

File: apps/contributions/tests/test_views.py (hunk 4 above)
Comment (x110dc) [reply apps/contributions/tests/test_views.py]: I was wondering about that. Thanks!

File: apps/contributions/views.py (hunk 3 above)
Comment (x110dc) [reply apps/contributions/views.py]: I changed it to "Error"

File: apps/contributions/tests/test_views.py (hunk 5 above)
Comment (x110dc) [reply apps/contributions/tests/test_views.py:790]: It's not meaningful. When I've exhausted "foo", "bar", "baz" and "quux" then I don't know what to use next so "fiddle" often comes to mind; but yes, I can see the value in adding something descriptive there. Will change.

**Pull request merged**